tracing-tracy/src/lib.rs

```diff
@@ -228,37 +256,52 @@ where
     fn on_enter(&self, id: &Id, ctx: Context<S>) {
         let Some(span) = ctx.span(id) else { return };
 
+        let extensions = span.extensions();
+        let fields = extensions.get::<FormattedFields<F>>();
         let stack_frame = {
             let metadata = span.metadata();
-            let name: Cow<str> = if let Some(fields) = span.extensions().get::<FormattedFields<F>>()
-            {
-                if fields.fields.is_empty() {
-                    metadata.name().into()
-                } else {
-                    format!("{}{{{}}}", metadata.name(), fields.fields.as_str()).into()
-                }
-            } else {
-                metadata.name().into()
-            };
-
             let file = metadata.file().unwrap_or("<not available>");
             let line = metadata.line().unwrap_or(0);
-            (
-                self.client.clone().span_alloc(
-                    Some(self.truncate_to_length(
-                        &name,
-                        file,
+            let span = |name: &str| {
+                (
+                    self.client.clone().span_alloc(
+                        Some(self.truncate_span_to_length(
+                            name,
+                            file,
+                            "",
+                            "span information is too long and was truncated",
+                        )),
                         "",
-                        "span information is too long and was truncated",
-                    )),
-                    "",
-                    file,
-                    line,
-                    self.stack_depth,
-                ),
-                id.into_u64(),
-            )
+                        file,
+                        line,
+                        self.stack_depth,
+                    ),
+                    id.into_u64(),
+                )
+            };
+
+            if let Some(fields) = self
+                .fields_in_zone_name
+                .then_some(fields)
+                .flatten()
+                .filter(|f| !f.fields.is_empty())
+            {
+                CACHE.with(|cache| {
+                    let mut buf = cache.acquire();
+                    let _ = write!(buf, "{}{{{}}}", metadata.name(), fields.fields);
+                    span(&buf)
+                })
+            } else {
+                span(metadata.name())
+            }
         };
+        if let Some(fields) = (!self.fields_in_zone_name).then_some(fields).flatten() {
+            stack_frame.0.emit_text(self.truncate_to_length(
+                (u16::MAX - 1).into(),
+                &fields.fields,
+                "span field values are too long and were truncated",
+            ));
+        }
```

Comment:
FWIW `then_some(()).and(fields)` might have been nicer and avoids constructing doubly-nested option.

Alternatively, you could write

```suggestion
        match fields {
            None => span(metadata.name()),
            // NOTE: the current proposed code will `emit_text` even for empty fields!
            Some(fields) if fields.is_empty() => span(metadata.name()),
            Some(fields) if self.fields_in_zone_name => {
                CACHE.with(|cache| {
                    let mut buf = cache.acquire();
                    let _ = write!(buf, "{}{{{}}}", metadata.name(), fields.fields);
                    span(&buf)
                })
            }
            Some(fields) => {
                let span = span(metadata.name());
                span.emit_text(...);
                span
            }
        }
```

which seems a more straightforward read, isn’t it?

Reply:
That's quite a bit clearer, thanks!